oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/OAuth2AuthorizationGrant.java

```diff
@@ -42,12 +43,12 @@
      * @return {@link CompletionStage} that refers to {@link HttpRequest} wrapped wrap with
      *         OAuth 2.0 authorization information.
      */
-    default CompletionStage<HttpRequest> withAuthorization(HttpRequest req) {
-        return getAccessToken().thenApply(accessToken -> {
-            // Create a new request with an additional 'Authorization' header
-            return req.withHeaders(req.headers().toBuilder()
-                                   .set(HttpHeaderNames.AUTHORIZATION, accessToken.authorization())
-                                   .build());
+    default CompletionStage<HttpRequest> withAuthorization(ClientRequestContext ctx, HttpRequest req) {
```

Comment:
I think users could call this method before a `ClientRequestContext` is made.
If you want to acquire an event loop, how about forking the following code?
https://github.com/line/armeria/blob/73c4e0ffcc0a544ee2b435aefd27acf973fece03/core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java#L442-L443

Reply:
Thanks for the advice!

My understanding was `XxxGrant` classes are only meant to be used in `OAuth2Client` decorator.
So I've thought we need a `ContextAwareEventLoop` as the authentication occurs within the context of armeria's client request.

Maybe I am not getting the whole picture of this implementation 😂

Reply:
> My understanding was XxxGrant classes are only meant to be used in OAuth2Client decorator.

Acutally, `OAuth2AuthorizationGrant.withAuthorization(req)` was suggested by me to inject `Authorization` header into the existing `HttpRequest`.

Reply:
Thanks for the reference, let me check the concept and update the PR.

Reply:
I am sorry for not being able to update the PR in a timely manner.
I've reverted a change in the interface as per your suggestion and the approach suggested by @minwoox removed the needs for acquiring an event loop in the end.

But, I still think being able to access `ctx` in `withAuthorization` may benefit us in a few situations.

Perhaps, we may want to introduce a new grant type in the future like `TokenPerUserXXXGrantType` where we need to load an access token for each logged in user and `ctx` will most likely be the right place where we put or get`User` objects. Then, it's just a matter of overriding `abstract User loadUser(ClientRequestContext ctx)` method.

Reply:
Hm, I don't see any case that a user would call `withAuthorization` before a `RequestContext` is created. Let me talk to @ikhoon later. 😄

Reply:
I imagine that users could want to acquire an access token using an `OAuth2ClientCredentialsGrant` and send a request to another resource server.
However, I am not an expert in OAuth2. Any thoughts @max904-github?

Reply:
We may handle this in another PR once we got an answer from @max904-github!

Reply:
Oops, I forgot to mention about it. 😅 Let me talk to @ikhoon.

Reply:
Had a chat with @ikhoon about it.
We can hide this API `withAuthorization` because it's only used in `OAuth2Client`.
How about moving the method inside `OAuth2Client` and make it private?

Reply:
@minwoox Thanks for the suggestion!
I've moved `withAuthorization` method to `OAuth2Client` but still without modifying the method signature.
Please check out if this is what you'd expect. 🙏

Reply:
How about just inlining it so that we don't create an additional future? 😄

Reply:
I shouldn't have just copy & pasted 😂.
All updated!